Sprint_8

Walkthroughcoding opened the pull request:
- Fix epic status update logic to properly reflect subtask statuses
- Ensure epics transition to IN_PROGRESS when subtasks have mixed statuses
- Refactor subtask addition and deletion to maintain consistency
- Improve test coverage and validation for epic-subtask relationships

EugeneLenkevich commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -37,9 +35,15 @@ private int generateId() { // Исправлено на private
+            prioritizedTasks.add(task);
+            System.out.println("Задача с id " + id + " добавлена.");
+        } else {
+            throw new IllegalArgumentException("Ошибка: Задача пересекается по времени с другой задачей.");
```
IllegalArgumentException

Ну не, это лучше не использовать, оно означает другое - это некоторая ошибка в самом алгоритме, давай лучше сделаем свое _проверяемое_ исключение (extends Exception). Так мы обозначим конкретную ситуацию конкретно в нашем бизнес-процессе :)

EugeneLenkevich commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -194,28 +207,36 @@ public Task getAnyTask(int id) {
             System.out.println("Подзадача с id " + id + " удалена.");
         } else {
-            System.out.println("Подзадача с id " + id + " не найдена.");
+            System.out.println("Ошибка: Подзадача с id " + id + " не найдена.");
```
Лучше кинуть проверяемое исключение, вдруг наш сервис вызывает другой сервис - он в лог посмотреть не сможет, а так узнает о ситуации и примет какие то действия в зависимости от своей цели, с которой вызывал наш сервис

EugeneLenkevich reviewed (changes requested):
Привет, Егор, спасибо за присланное решение

Отличная работа! :)
Есть совсем небольшие замечания, посмотри комментарии, пожалуйста
Пару моментиков с исключениями

И про исключения вот такое объяснение, на всякий случай добавлю тут.

Есть еще такой признак - если исключение есть в бизнес-процессе, то его стоит сделать проверяемым. Так мы на уровне компилятора говорим о возможных альтернативных ситуациях и обязываем вышестоящий  обработать их или эскалировать еще выше. По большому счету это и не ошибки вовсе, а некоторые отклонения от основного пути пользователя.
Например, снимаем денежку с карты в банкомате
стандартный путь: приложили карту, получили денежку
возможные отклонения
- ввели сумму больше чем денег на балансе
- карта заблокирована 
- ввели неверный пин-код
- и пр

пример ошибки: у банкомата пропало соединение с интернетом (может и так он будет работать, не в курсе, но пусть будет что не может), эту ситуацию обрабатывать бессмысленно кроме как отобразить пользователю сообщение что "атм не работает" и прекратить все текущие операции до устранения внешней причины. И в бизнес-процессе этой ситуации нет, это чисто техническая история

Walkthroughcoding committed:
- Fix epic status update logic to properly reflect subtask statuses
- Ensure epics transition to IN_PROGRESS when subtasks have mixed statuses
- Refactor subtask addition and deletion to maintain consistency
- Improve test coverage and validation for epic-subtask relationships

Walkthroughcoding committed:
fix: исправление ошибок Checkstyle (удалены неиспользуемые импорты)

Walkthroughcoding committed:
fix: исправлены тесты epicStatusUpdatesWithSubtasks() и addNewEpicWithSubtasks()

Walkthroughcoding committed:
- Теперь при обнаружении конфликта времени выбрасывается `TaskTimeConflictException`.
- Обновлены тесты для проверки корректной работы исключений.

EugeneLenkevich reviewed (approved):
Теперь все отлично, спасибо :)

Walkthroughcoding committed:
- Исправлен баг с методом isOverlapping()

Walkthroughcoding committed:
- Скорректирована обработка исключений в тесте, в интервейс TaskManager добавлена обработка TaskNotFoundException
